crates/bevy_render/src/color/mod.rs

```diff
@@ -1332,38 +1322,38 @@ impl encase::private::CreateFrom for Color {
 
 impl encase::ShaderSize for Color {}
 
-#[derive(Debug, Error)]
+#[derive(Debug, Error, PartialEq, Eq)]
 pub enum HexColorError {
     #[error("Unexpected length of hex string")]
     Length,
-    #[error("Error parsing hex value")]
-    Hex(#[from] hex::FromHexError),
+    #[error("Invalid hex char")]
+    Char(char),
 }
 
-fn decode_rgb(data: &[u8]) -> Result<Color, HexColorError> {
-    let mut buf = [0; 3];
-    match hex::decode_to_slice(data, &mut buf) {
-        Ok(_) => {
-            let r = buf[0] as f32 / 255.0;
-            let g = buf[1] as f32 / 255.0;
-            let b = buf[2] as f32 / 255.0;
-            Ok(Color::rgb(r, g, b))
-        }
-        Err(err) => Err(HexColorError::Hex(err)),
+const fn decode_hex<const N: usize>(mut bytes: [u8; N]) -> Result<[u8; N], HexColorError> {
+    let mut i = 0;
+    while i < bytes.len() {
+        let val = match hex_ascii_byte(bytes[i]) {
```

Comment:
This could also use `?`  syntax now.

Reply:
`?` is not allowed in a `const fn`. note: see issue #74935 <https://github.com/rust-lang/rust/issues/74935> for more information

Reply:
Similarly, for loops currently aren't allowed in `const fn`s, hence the while loops. (see https://github.com/rust-lang/rust/issues/87575 for info)

It would be really cool if the `const_generic_exprs` feature became mature enough to write

```rust
const fn decode_hex<const N: usize>(mut bytes: [u8; N]) -> Result<[u8; N / 2], HexColorError>
```

(Not that `N` would be so large in this context so as to be problematic, but I wonder if such a signature would necessarily require a new array to be constructed. I guess it might not be such a positive then.)